### ISSUE TITLE
How to access values from a custom item

### ISSUE THREAD

**Issue by danielsaidi** [danielsaidi/Sheeeeeeeeet#52]
I received this e-mail:

Good afternoon!
 
I have a question about your Sheeeeeeeeet Cocoapod. If I create a custom view with things such as a UITextField, DatePicker, etc, how would I go about accessing these values when I click OK? Here is some code to show what I was attempting to do:
 
```swift
let title = ActionSheetTitle(title: "App Info")
let customType = CustomViewCell.self
let item1 = ActionSheetCustomItem(cellType: customType) { cell in
let formatter = DateFormatter()
formatter.dateFormat = "MM/dd/yyyy"
            
       //set to current date
cell.updatedLabel.text = formatter.string(from: Date())
}
       
let button = ActionSheetOkButton(title: "Save")
return ActionSheet(items: [title, item1, button]) { sheet, item in
   
if item is ActionSheetOkButton {
             print("Save buttons has the value `true`")
              //access the values in item1 and do something with the data
       }
}
```

Thoughts on what I am missing?

**Comment by danielsaidi** [comment]
Hi David!

The custom item type is rather new, so I haven't polished it much after releasing it.

The custom item is supposed to be self contained, so you can model it in any way you like. However, this also means that it is not (yet) as straightforward and obvious as other item types.

I mostly use the custom item type with cells that have action blocks that I set from the outside. This means that if I open a sheet with a custom item, it's cell usually triggers actions that do something outside of the sheet, e.g. modify the current view controller. However, this is not the normal use, since these cells do not update a model in the sheet.

The custom item currently has no reference to the cell, which means that you cannot access the cell in the select action, even if you get a reference to your item as such:

```swift
return ActionSheet(items: [title, item1, button]) { sheet, item in
   let item = sheet.items.compactMap { $0 as? ActionSheetCustomItem }.first
   ...
}
```

However, until I improve the custom item type, you could try these approaches:

* Subclass `ActionSheetCustomItem` and add a model to it, e.g. selected date, entered text etc. Then have the cell update that model. You could then cast as above and access the model.
* Subclass the `ActionSheet` and add a model to it, then do the same as above, but update the sheet instead of the item.

I hope this makes sense. Please let me know if this works.

**Comment by tfworld** [comment]
Thank you for looking into this! I came up with a pretty terrible way to access my items... It involves going through the view stack until I find the correct class. My next step will be to see if I can add a reference to the custom class in the sheet object so that I can directly access my fields. Here is the UI I made for this use case.

![alt text](http://tfworld.com/images/info.png "Info")

**Comment by danielsaidi** [comment]
Thank you for filling me in with more info (and a nice screenshot! :) )

Ok, sounds like you have a plan! I will see if I can improve the demo app with more custom item types that handle actions and selections differently. 

Just let me know if you come up with any ideas on how to improve this item type. 👍

**Comment by danielsaidi** [comment]
This has been living for a long time, but I will close it without action. I experimented tonight with adding another action (an anonymous "callback" action) to the custom item. It worked, but it was in no way an improvement. It just resulted in messier code. I think you'll just have to inject something into the already existing setup action, since my approach ended up doing that anyway.

Closing this for now.
